src/ray/core_worker/experimental_mutable_object_manager.h

```diff
@@ -115,38 +115,43 @@ class ExperimentalMutableObjectManager {
 
  private:
   struct WriterChannel {
+#ifdef __linux__
     WriterChannel(std::unique_ptr<plasma::MutableObject> mutable_object_ptr)
         : mutable_object(std::move(mutable_object_ptr)) {}
 
     /// Whether the object is sealed, i.e. whether the last writer still needs
     /// to call WriteRelease.
     bool is_sealed = true;
     std::unique_ptr<plasma::MutableObject> mutable_object;
+#endif
   };
 
+  // Thread-safe for multiple ReadAcquire threads and one ReadRelease thread.
   struct ReaderChannel {
+#ifdef __linux__
     ReaderChannel(std::unique_ptr<plasma::MutableObject> mutable_object_ptr)
-        : mutable_object(std::move(mutable_object_ptr)) {}
+        : mutable_object(std::move(mutable_object_ptr)) {
+      RAY_CHECK(sem_init(&reader_semaphore, /*pshared=*/0, /*value=*/1) == 0);
+    }
 
     /// The last version that we read. To read again, we must pass a newer
     /// version than this.
     int64_t next_version_to_read = 1;
+    /// Use a semaphore to protect next_version_to_read. This is necessary if
+    /// begin_read and end_read are called from different threads.
+    sem_t reader_semaphore;
     /// Whether we currently have a read lock on the object. If this is true,
-    /// then it is safe to read the value of the object. For immutable objects,
-    /// this will always be true once the object has been sealed. For mutable
-    /// objects, ReadRelease resets this to false, and ReadAcquire resets to
-    /// true.
+    /// then it is safe to read the value of the object.
     bool read_acquired = false;
     std::unique_ptr<plasma::MutableObject> mutable_object;
+#endif
   };
 
-  Status EnsureGetAcquired(ReaderChannel &channel);
-
-  /// All channels for which we are registered as a writer. This can overlap
-  /// with reader channels (e.g., if the CoreWorker is multithreaded and one
-  /// thread reads while the other writes).
+  // TODO(swang): Access to these maps is not threadsafe. This is fine in the
```

Comment:
Nit; To be safer, we can create Add/Remove Channel APi with FATAL error? And we can add a comment there?